### PR TITLE
chore: add 'distribute' GitHub Action containing build/test/distribute steps

### PR DESCRIPTION
**Pull request by jijiechen**
### Checklist prior to review

- [ ] [Link to relevant issue][1] as well as docs and UI issues --
  - not applicable
- [x] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [ ] Tests (Unit test, E2E tests, manual test on universal and k8s)
  - not applicable
  - Don't forget `ci/` labels to run additional/fewer tests
- [ ] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? 
  - no need
- [ ] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) 
  - not applicable

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

A WIP effort for setting up commit CI based on GitHub Actions.

### CODE REVIEW

File: .github/workflows/distribute.yaml [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+          esac
+      # GitHub actions does not share cache across multiple jobs,
+      # so we have to operate cache in each job and action file
+      - id: retrieve-cache
+        uses: actions/cache/restore@v3
+        with:
+          path: |
+            ~/.cache/go-build
+            ~/.kuma-dev
+          key: ${{ runner.os }}-build-${{ hashFiles('**/go.sum') }}-${{ hashFiles('mk/dependencies/deps.lock') }}
```
Comment (lahabana) [inline .github/workflows/distribute.yaml:38]: Remove the `~/.cache/go-build` (it's done by go already).

(Move the go action above this).

Remove from the cache Key `${{ hashFiles('**/go.sum') }}` (there's also an option to use the "closest" cache in github)

File: .github/workflows/distribute.yaml [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+      - uses: actions/setup-go@v4
+        with:
+          go-version-file: go.mod
+      - if: steps.retrieve-cache.outputs.cache-hit != 'true'
```
Comment (lahabana) [inline .github/workflows/distribute.yaml:42]:
```suggestion
```

Remove this `make dev/tools` should be almost a noop if the cache is up to date.

File: .github/workflows/distribute.yaml [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+              apt install -y sudo
+            fi
+            sudo apt update
+            sudo env DEBIAN_FRONTEND=noninteractive apt install -y curl git make unzip gcc xz-utils
+            ;;
+          esac
+      # GitHub actions does not share cache across multiple jobs,
+      # so we have to operate cache in each job and action file
+      - id: retrieve-cache
+        uses: actions/cache/restore@v3
```
Comment (lahabana) [inline .github/workflows/distribute.yaml:33]: Just use `actions/cache` instead of both restore and save.

File: .github/workflows/distribute.yaml [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+          case "$ID" in
+          ubuntu)
+            if ! command -v sudo 2>&1 >/dev/null; then
+              apt update
+              apt install -y sudo
+            fi
+            sudo apt update
+            sudo env DEBIAN_FRONTEND=noninteractive apt install -y curl git make unzip gcc xz-utils
+            ;;
+          esac
```
Comment (lahabana) [inline .github/workflows/distribute.yaml:93]: Might want to remove this.

File: .github/workflows/distribute.yaml [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+          path: build
+          retention-days: 1
+  test:
+    needs: build
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+          ref: ${{ github.event_name == 'pull_request' && github.event.pull_request.head.sha || github.sha }}
```
Comment (lahabana) [inline .github/workflows/distribute.yaml:77]: I think you don't need the ref.

File: .github/workflows/distribute.yaml [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        uses: actions/upload-artifact@v2
+        with:
+          name: build-output
+          path: build
+          retention-days: 1
+  test:
+    needs: build
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
```
Comment (lahabana) [inline .github/workflows/distribute.yaml:74]: Add option to skip this in `ci/skip-test` is set as a label

**Review by lahabana — COMMENTED** [review]
It's starting to take great shape! Let me know once you've integrated the few changes and if it passes on your fork I'll merge it :)

File: .github/workflows/distribute.yaml [hunk 7]
```diff
@@ -0,0 +1,162 @@
+name: "distribute"
```
Comment (lahabana) [inline .github/workflows/distribute.yaml:1]:
```suggestion
name: "build-test-distribute"
```

**Comment by jijiechen** [comment]
I'll open a new Pull Request with some new features (), and I'll also merge all the suggested changes in this thread into that new one. @lahabana

File: .github/workflows/distribute.yaml [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+          case "$ID" in
+          ubuntu)
+            if ! command -v sudo 2>&1 >/dev/null; then
+              apt update
+              apt install -y sudo
+            fi
+            sudo apt update
+            sudo env DEBIAN_FRONTEND=noninteractive apt install -y curl git make unzip gcc xz-utils
+            ;;
+          esac
```
[the remark this answers is not in the document]
Comment (jijiechen) [reply .github/workflows/distribute.yaml:28]: I'll check if it works without these.

File: .github/workflows/distribute.yaml [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+        run: |
+          rm -rf ./build
+      - uses: actions/download-artifact@v2
+        with:
+          name: build-output
+          path: build
+      - run: |
+          sudo apt-get update; sudo apt-get install -y qemu-user-static binfmt-support
+      - run: |
+          make test/container-structure
```
[the remark this answers is not in the document]
Comment (jijiechen) [reply .github/workflows/distribute.yaml:120]: Is this task a test? In my understanding, the `distribute` task is make distributions (as the name implies), and is not intended to test/check.

File: .github/workflows/distribute.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (jijiechen) [reply .github/workflows/distribute.yaml:28]: Confirmed, it works. So I'll remove this.

**Pull request closed without merging**